File: unitytogradle/src/main/java/com/commit451/unitytogradle/UnityProject.java
package com.commit451.unitytogradle;

import java.io.File;
import java.io.IOException;

/**
 * Defines the structure of the current Unity project
 */
public class UnityProject {

    public static UnityProject from(String pathToRootOfProject) throws IOException {
        UnityProject unityProject = new UnityProject();
        unityProject.project = new File(pathToRootOfProject);
        unityProject.assets = new File(unityProject.project, "assets");
        unityProject.libs = new File(unityProject.project, "libs");
        unityProject.res = new File(unityProject.project, "res");
        unityProject.src = new File(unityProject.project, "src");
        unityProject.manifest = new File(unityProject.project, "AndroidManifest.xml");
        unityProject.proguard = new File(unityProject.project, "proguard-unity.txt");

        String fileContent = Utils.loadFileAsString(unityProject.manifest);
        unityProject.packageName = parsePackageNameFromManifest(fileContent);
        unityProject.versionName = parseVersionNameFromManifest(fileContent);
        unityProject.versionCode = parseVersionCodeFromManifest(fileContent);
        unityProject.minSdkVersion = parseMinSdkVersion(fileContent);
        unityProject.targetSdkVersion = parseTargetSdkVersion(fileContent);
        return unityProject;
    }

    private static String parsePackageNameFromManifest(String fileContent) throws IOException {
        String text = "package=\"";
        int indexOfStart = fileContent.indexOf(text) + text.length();
        int inextOfEnd = fileContent.indexOf("\"", indexOfStart);
        return fileContent.substring(indexOfStart, inextOfEnd);
    }

    private static String parseVersionNameFromManifest(String fileContent) throws IOException {
        String text = "versionName=\"";
        int indexOfStart = fileContent.indexOf(text) + text.length();
        int inextOfEnd = fileContent.indexOf("\"", indexOfStart);
        return fileContent.substring(indexOfStart, inextOfEnd);
    }

    private static String parseVersionCodeFromManifest(String fileContent) throws IOException {
        String text = "versionCode=\"";
        int indexOfStart = fileContent.indexOf(text) + text.length();
        int inextOfEnd = fileContent.indexOf("\"", indexOfStart);
        return fileContent.substring(indexOfStart, inextOfEnd);
    }

    private static String parseMinSdkVersion(String fileContent) throws IOException {
        String text = "minSdkVersion=\"";
        int indexOfStart = fileContent.indexOf(text) + text.length();
        int inextOfEnd = fileContent.indexOf("\"", indexOfStart);
        return fileContent.substring(indexOfStart, inextOfEnd);
    }

    private static String parseTargetSdkVersion(String fileContent) throws IOException {
        String text = "targetSdkVersion=\"";
        int indexOfStart = fileContent.indexOf(text) + text.length();
        int inextOfEnd = fileContent.indexOf("\"", indexOfStart);
        return fileContent.substring(indexOfStart, inextOfEnd);
    }

    public File project;
    public File assets;
    public File libs;
    public File res;
    public File src;
    public File manifest;
    public File proguard;

    public String packageName;
    public String versionName;
    public String versionCode;
    public String minSdkVersion;
    public String targetSdkVersion;

    private UnityProject() {
    }


}

File: unitytogradle/src/main/java/com/commit451/unitytogradle/UnityProject.kt
package com.commit451.unitytogradle

import java.io.File

/**
 * Defines the structure of the current Unity project
 */
class UnityProject private constructor() {

    lateinit var project: File
    lateinit var assets: File
    lateinit var libs: File
    lateinit var res: File
    lateinit var src: File
    lateinit var manifest: File
    lateinit var proguard: File

    lateinit var packageName: String
    lateinit var versionName: String
    lateinit var versionCode: String
    lateinit var minSdkVersion: String
    lateinit var targetSdkVersion: String

    companion object {

        fun from(pathToRootOfProject: String): UnityProject {
            val unityProject = UnityProject()
            unityProject.project = File(pathToRootOfProject)
            unityProject.assets = File(unityProject.project, "assets")
            unityProject.libs = File(unityProject.project, "libs")
            unityProject.res = File(unityProject.project, "res")
            unityProject.src = File(unityProject.project, "src")
            unityProject.manifest = File(unityProject.project, "AndroidManifest.xml")
            unityProject.proguard = File(unityProject.project, "proguard-unity.txt")

            val fileContent = Utils.loadFileAsString(unityProject.manifest)
            unityProject.packageName = parsePackageNameFromManifest(fileContent)
            unityProject.versionName = parseVersionNameFromManifest(fileContent)
            unityProject.versionCode = parseVersionCodeFromManifest(fileContent)
            unityProject.minSdkVersion = parseMinSdkVersion(fileContent)
            unityProject.targetSdkVersion = parseTargetSdkVersion(fileContent)
            return unityProject
        }

        private fun parsePackageNameFromManifest(fileContent: String): String {
            val text = "package=\""
            val indexOfStart = fileContent.indexOf(text) + text.length
            val inextOfEnd = fileContent.indexOf("\"", indexOfStart)
            return fileContent.substring(indexOfStart, inextOfEnd)
        }

        private fun parseVersionNameFromManifest(fileContent: String): String {
            val text = "versionName=\""
            val indexOfStart = fileContent.indexOf(text) + text.length
            val inextOfEnd = fileContent.indexOf("\"", indexOfStart)
            return fileContent.substring(indexOfStart, inextOfEnd)
        }

        private fun parseVersionCodeFromManifest(fileContent: String): String {
            val text = "versionCode=\""
            val indexOfStart = fileContent.indexOf(text) + text.length
            val inextOfEnd = fileContent.indexOf("\"", indexOfStart)
            return fileContent.substring(indexOfStart, inextOfEnd)
        }

        private fun parseMinSdkVersion(fileContent: String): String {
            val text = "minSdkVersion=\""
            val indexOfStart = fileContent.indexOf(text) + text.length
            val inextOfEnd = fileContent.indexOf("\"", indexOfStart)
            return fileContent.substring(indexOfStart, inextOfEnd)
        }

        private fun parseTargetSdkVersion(fileContent: String): String {
            val text = "targetSdkVersion=\""
            val indexOfStart = fileContent.indexOf(text) + text.length
            val inextOfEnd = fileContent.indexOf("\"", indexOfStart)
            return fileContent.substring(indexOfStart, inextOfEnd)
        }
    }


}
